// typescript-functional-extensions/dist/cjs/utilities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.noop = exports.pipeFromArray = exports.isPromise = exports.isFunction = exports.isNone = exports.isSome = exports.isDefined = exports.never = void 0;
const never = () => {
    throw Error('This error should be unreachable');
};
exports.never = never;
function isDefined(value) {
    return value !== undefined && value !== null;
}
exports.isDefined = isDefined;
function isSome(value) {
    return value !== undefined && value !== null;
}
exports.isSome = isSome;
function isNone(value) {
    return !isDefined(value);
}
exports.isNone = isNone;
function isFunction(value) {
    return typeof value === 'function';
}
exports.isFunction = isFunction;
function isPromise(value) {
    return value instanceof Promise;
}
exports.isPromise = isPromise;
/**
 * Sourced from https://github.com/ReactiveX/rxjs/blob/7268bd31d1cb30cf01a1a69a7b14458e15b76b58/src/internal/util/pipe.ts
 * @param fns
 * @returns
 */
function pipeFromArray(fns) {
    if (fns.length === 0) {
        return noop;
    }
    if (fns.length === 1) {
        return fns[0];
    }
    return function piped(input) {
        return fns.reduce((prev, fn) => fn(prev), input);
    };
}
exports.pipeFromArray = pipeFromArray;
function noop(value) {
    if (value) {
        return value;
    }
}
exports.noop = noop;
